### PR TITLE
feat: Job String Processor

### PR DESCRIPTION
**Pull request by merlin-northern**
Changelog: Title
Ticket: MEN-5819
Signed-off-by: Peter Grzybowski <peter@northern.tech>

### CODE REVIEW

**Comment by merlin-northern** [comment]
@mender-test-bot start pipeline please

**Comment by merlin-northern** [comment]
@mender-test-bot start pipeline please

**Comment by merlin-northern** [comment]
@mender-test-bot start pipeline please

**Comment by merlin-northern** [comment]
@tranchitella could I kindly ask you to take a look, since you already reviewed the first try (the quck fix from the past weekend in ent)

File: app/worker/process.go [hunk 1]
```diff
@@ -142,25 +144,23 @@ func processTask(task model.Task, job *model.Job,
 				"Error: Task definition incompatible " +
 					"with specified type (http)")
 		}
-		l.Infof("processTask: calling http task: %s %s", httpTask.Method,
-			processJobString(httpTask.URI, workflow, job))
```
Comment (tranchitella) [inline app/worker/process.go:146]: Why did you remove this log?

File: app/worker/http.go [hunk 2]
```diff
@@ -39,33 +40,42 @@ var makeHTTPRequest = func(req *http.Request, timeout time.Duration) (*http.Resp
...
+	ps *processor.JobStringProcessor,
+	l *log.Logger,
+) (*model.TaskResult, error) {
+	options := &processor.Options{Encoding: processor.URL}
+	uri := ps.ProcessJobString(httpTask.URI, options)
+	l.Infof("processHTTPTask: starting with: method=%s uri=%s (options=%+v)",
+		httpTask.Method,
+		uri,
+		options,
+	)
```
Comment (tranchitella) [inline app/worker/http.go:54]: I don't think logging option is valuable, as it will always be Encoding: URL.
By the way, I don't get why we need this log here.

File: app/worker/http.go [hunk 3]
```diff
@@ -39,33 +40,42 @@ var makeHTTPRequest = func(req *http.Request, timeout time.Duration) (*http.Resp
...
 	var payloadString string
 	if len(httpTask.FormData) > 0 {
 		form := url.Values{}
 		for key, value := range httpTask.FormData {
-			key = processJobString(key, workflow, job)
-			key = maybeExecuteGoTemplate(key, job.InputParameters.Map())
-			value = processJobString(value, workflow, job)
-			value = maybeExecuteGoTemplate(value, job.InputParameters.Map())
+			key = ps.ProcessJobString(key, nil)
+			key = maybeExecuteGoTemplate(key, ps.GetJobInputParameters().Map())
```
Comment (tranchitella) [inline app/worker/http.go:61]: It looks weird obtaining the job's input parameters from a `GetJobInputParameters` method on the JobStringProcessor instance. It looks like you did it by convenience because processHTTPTask doesn't receive anymore the job, but honestly I don't like it too much.

File: app/worker/process.go (hunk 1 above)
Comment (tranchitella) [inline app/worker/process.go:146]: why did you remove this log and moved it in HTTPTask?

File: app/processor/string.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+	workflow *model.Workflow,
+	job *model.Job,
+) *JobStringProcessor {
+	return &JobStringProcessor{
+		workflow: workflow,
+		job:      job,
+	}
+}
+
+func (j *JobStringProcessor) ProcessJobString(data string, options *Options) string {
```
Comment (tranchitella) [inline app/processor/string.go:65]: If you use a variadic function parameter, you can save the `, nil` where you don't need otpions:

```suggestion
func (j *JobStringProcessor) ProcessJobString(data string, opts ...*Options) string {
```

At this point, you can loop over options (not tested):
```
    options := Options{Encoding: Plain}
    for _, o := range opts {
        options.Encoding = o.Encoding
    }
```

If you want to spend a bit more time, you can use the functional options pattern:
https://www.sohamkamani.com/golang/options-pattern/

File: app/worker/process.go (hunk 1 above)
Comment (merlin-northern) [reply app/worker/process.go:146]: 1. I moved it below
2. it was doubling the very expensive operation of processing the job string, just for printing.

File: app/worker/http.go (hunk 2 above)
Comment (merlin-northern) [reply app/worker/http.go:54]: I can remove it, it was moved from one level up (see the previous comment). do you want me to remove it?

File: app/worker/process.go (hunk 1 above)
Comment (merlin-northern) [reply app/worker/process.go:146]: it was super expensive to process the job string twice, here for only logging and then again right after.

File: app/worker/http.go (hunk 3 above)
Comment (merlin-northern) [reply app/worker/http.go:61]: I know, but there is no good way of going it. would you prefer `ps.GetJob()` method instead?

File: app/processor/string.go (hunk 4 above)
Comment (merlin-northern) [reply app/processor/string.go]: I was following your previous suggestions :) I like the `...`. the functional options: nice! I have never used it. I think `...` is nice here. is it ok now?

File: app/worker/http.go (hunk 3 above)
Comment (merlin-northern) [reply app/worker/http.go:61]: @alfrunes there is a bit of refactoring going on here (I could not stop) in order not to redesign the whole thing, I added `GetJobInputParameters` in the string processor class. we (Fabio and me) both think it is not very elegant, and a bit out of place. what do you think? (Keep in mind that we do not want to go over board here :>)

Comment (alfrunes) [reply app/worker/http.go:61]: Since `maybeExecuteGoTemplate` is called wherever `ps.ProcessJobString` is called; wouldn't it be easier to just move it inside the `processor` package?

File: app/processor/string.go (hunk 4 above)
Comment (merlin-northern) [reply app/processor/string.go]: will you believe it? `app/processor/string.go:67:1: cyclomatic complexity 21 of func `(*JobStringProcessor).ProcessJobString` is high (> 20) (gocyclo)` there is always something.

File: app/worker/http.go (hunk 3 above)
Comment (merlin-northern) [reply app/worker/http.go]: brilliant. so I created a method `MaybeExecuteGoTemplate` but there were complications, in the end, this is the final version. makes sense? beware: I can refactor deeper.

**Comment by merlin-northern** [comment]
@mender-test-bot start pipeline please sugar

**Comment by alfrunes** [comment]
Is this a `feat`ure? I would argue that this is a `refac`tor of existing functionality.

File: app/processor/job.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+	"github.com/mendersoftware/workflows/model"
+)
+
+type JobProcessor struct {
+	job *model.Job
+}
+
+type JsonOptions struct {
+}
```
Comment (alfrunes) [inline app/processor/job.go:30]: I don't understand the purpose of this.

File: app/processor/job.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+func NewJobProcessor(job *model.Job) *JobProcessor {
+	return &JobProcessor{
+		job: job,
+	}
+}
+
+func (j JobProcessor) ProcessJSON(
+	data interface{},
+	ps *JobStringProcessor,
+	options ...*JsonOptions,
```
Comment (alfrunes) [inline app/processor/job.go:41]: This variadic argument is unused.

File: app/processor/job.go (hunk 5 above)
Comment (merlin-northern) [reply app/processor/job.go:30]: this is on purpose: to keep the interface consistent. in case there is a new round of refactoring (i.e.: a general processor interface) can we keep it?

File: app/processor/job.go (hunk 6 above)
Comment (merlin-northern) [reply app/processor/job.go:41]: this is on purpose: to keep the interface consistent. in case there is a new round of refactoring (i.e.: a general processor interface) can we keep it?

**Comment by merlin-northern** [comment]
> Is this a `feat`ure? I would argue that this is a `refac`tor of existing functionality.

something in between feat, refact, fix, I will change to fix, since it fixes a bug. makes sense?

File: app/worker/http.go [hunk 7]
```diff
@@ -39,34 +40,41 @@ var makeHTTPRequest = func(req *http.Request, timeout time.Duration) (*http.Resp
...
-	workflow *model.Workflow, l *log.Logger) (*model.TaskResult, error) {
-	uri := processJobString(httpTask.URI, workflow, job)
+func processHTTPTask(
+	httpTask *model.HTTPTask,
+	ps *processor.JobStringProcessor,
+	jp *processor.JobProcessor,
+	l *log.Logger,
+) (*model.TaskResult, error) {
+	options := &processor.Options{Encoding: processor.URL}
+	uri := ps.ProcessJobString(httpTask.URI, options)
```
Comment (alfrunes) [inline app/worker/http.go:50]: This is going to break certain URLs. For example if the url contains a `host:port` address. To be honest, I think the query escape should be done at the templating level and thus done as part of the workflow definition.

Comment (merlin-northern) [reply app/worker/http.go:50]: wait how can it be? it only URL-encodes the value:

[app/processor/string.go:97](https://github.com/mendersoftware/workflows/pull/234/files#diff-a395ff9022c3014278f32fad3c86b421391bc1558be6cf5ff6539e33cbbcbd89R97)
```go
						value = url.QueryEscape(value)
```

Comment (alfrunes) [reply app/worker/http.go:50]: This transforms all of the following characters: `'$', '&', '+', ',', '/', ':', ';', '=', '?', '@'` many of them are valid in different parts of the schema. For example
```go
url.QueryEscape("https://hosted.mender.io") // -> https%3A%2F%2Fhosted.mender.io
```

Comment (merlin-northern) [reply app/worker/http.go:50]: but the URL is never a workflow input variable, we take it only from env and with default values (`"uri": "http://${env.USERADM_ADDR|mender-useradm:8080}/api/internal/v1/useradm/tokens?tenant_id=${workflow.input.tenant_id}"`). this assumes that we encode the workflow input variables present in the URL (`${workflow.input.tenant_id}`).

**Comment by merlin-northern** [comment]
@mender-test-bot start pipeline please sugar pretty please

File: app/worker/http.go (hunk 7 above)
Comment (merlin-northern) [reply app/worker/http.go:50]: ok, how about now? we support cool `encoding=plain|url;identifier` syntax that overrides everything you may pass. if you do not pass it, by default we URL encode all input variables that exist in the URL.

Comment (merlin-northern) [reply app/worker/http.go:50]: @tranchitella I hope you are happy that you used Alf as a mole to have this feature :>

File: app/processor/string.go [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+			}
+		} else if strings.HasPrefix(match, workflowEnvVariable) &&
+			len(match) > len(workflowEnvVariable) {
+			// Replace ${env.KEY} with the KEY environment variable
+			envName := match[len(workflowEnvVariable):]
+			envValue := os.Getenv(envName)
+			if envValue == "" {
+				envValue = defaultValue
+			}
+			data = strings.ReplaceAll(data, submatch[0], envValue)
```
Comment (alfrunes) [inline app/processor/string.go:148]:
```suggestion
			if envValue := os.Getenv(envName); envValue != "" {
				value = envValue
			}
```

File: app/processor/string.go [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+					if varValue == nil {
+						varValue = ""
+					}
+					varValueString, err := utils.ConvertAnythingToString(varValue)
+					if err == nil {
+						if varValueString == "" {
+							varValueString = defaultValue
+						}
+						data = strings.ReplaceAll(data, submatch[0], varValueString)
+					}
```
Comment (alfrunes) [inline app/processor/string.go:174]:
```suggestion
					if err == nil {
						if varValueString != "" {
							value = varValueString
						}
					} else {
						continue SubmatchLoop
					}
```

File: app/processor/string.go [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+	}
+	return data, defaultEncoding
+}
+
+func (j *JobStringProcessor) ProcessJobString(data string, options ...*Options) string {
+	matches := reExpression.FindAllStringSubmatch(data, -1)
+	option := getOption(options...)
+
+	// search for ${...} expressions in the data string
+	for _, submatch := range matches {
```
Comment (alfrunes) [inline app/processor/string.go:113]:
```suggestion
SubmatchLoop:
	for _, submatch := range matches {
```

File: app/processor/string.go [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+					if err == nil {
+						if varValueString == "" {
+							varValueString = defaultValue
+						}
+						data = strings.ReplaceAll(data, submatch[0], varValueString)
+					}
+					break
+				}
+			}
+		}
```
Comment (alfrunes) [inline app/processor/string.go:178]: Finally, this applies the option to all the replacements above.
```suggestion
		}
		if option.Encoding == URL {
			value = url.QueryEscape(value)
		}
		data = strings.ReplaceAll(data, submatch[0], value)
```

**Comment by merlin-northern** [comment]
@alfrunes I moved to this PR: https://github.com/mendersoftware/workflows/pull/235 I fixed the units, and added tests for a new `processOptionString`, I had to rewrite that, ti was not working.

**Comment by merlin-northern** [comment]
this one went with https://github.com/mendersoftware/workflows/pull/235
